### DISCUSSION TITLE
Array Shuffle

### DISCUSSION THREAD

**Discussion by sbeneduce** [vlang/v#13793] [Ideas]
I use shuffle a lot and I use this function: i wonder if makes sense to add to the array lib

```v
import rand


pub fn shuffle<T>(a []T) []T {
	mut t := a.clone()
	mut s := []T{}

	
	for t.len > 0 {
		si := rand.intn(t.len) or {return a}
		s.prepend(t[si])
		t.delete(si)

	}	

	return s
}
```

**Comment by JalonSolov** [comment]
You won't get exactly what you want with that code.  You're prepending a value, then deleting what is now at the old position, but it's actually what was shifted down into that spot.

You need to save the value, delete it from the array, _then_ prepend it.

**Comment by ntrel** [comment]
I would write it in-place. Callers can call `clone` first if they need a copy.

```v
import rand

pub fn shuffle<T>(mut a []T) {
	for i in 0 .. a.len {
		si := rand.intn(a.len - i) or {return}
		t := a[si]
		a[si] = a[i]
		a[i] = t
	}	
}
```

**Reply by JalonSolov** [reply]
It works to just do
```v
a[i], a[si] = a[si], a[i]
```
instead of creating `t` to hold a value for a moment.
```v
import rand

pub fn shuffle<T>(mut a []T) {
        for i in 0 .. a.len {
                si := rand.intn(a.len - i) or {return}
                a[i], a[si] = a[si], a[i]
        }
}

fn main() {
        mut a := [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
        shuffle(mut a)
        println(a)
}
```
example output:
```
[9, 5, 2, 0, 6, 1, 4, 3, 7, 8]
```

**Reply by JalonSolov** [reply]
Although the rand may need to be adjusted, here... I just did it ten times in a row, and always wound up with `9` as the first element, while the rest changed every time.

**Reply by ntrel** [reply]
Thanks. Fixed and using your swap: https://github.com/vlang/v/pull/13811
